File: kotlin/153-find-minimum-in-rotated-sorted-array-Medium/Solution.kt
class Solution {
    fun findMin(nums: IntArray): Int {
        var l = 0
        var r = nums.lastIndex
        while (l < r) {
            val m = l + (r - l) / 2
            if (nums[m] > nums[r]) {
                l = m + 1
            } else {
                r = m
            }
        }
        return nums[l]
    }
}

File: scala/153-find-minimum-in-rotated-sorted-array-Medium/Solution.scala
object Solution {
    def findMin(nums: Array[Int]): Int = {
        def find(lo: Int, hi: Int): Int = {
            if (nums(lo) <= nums(hi))
                nums(lo)
            else {
                val pv = lo + (hi - lo) / 2
                if (nums(pv) < nums(hi))
                    find(lo, pv)
                else
                    find(pv + 1, hi)
            }
        }
        find(0, nums.length - 1)
    }
}